<a href="https://colab.research.google.com/github/priyanka011011/NLP/blob/main/model.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Use Case ZAF202305 Next Word Suggestion**

### **Overview**
We are building a web application with API access to a machine learning model without giving source of code or any details how it works, which helps you type faster on the laptop with a high quality next sword suggestions, based on what was typed so far. There are several possible scenarios:

### **Methodology**
The model is supposed to be trained on a large amount of texts. Suggested models are: LSTM Smoothed word Unigram models Smoothed word Bigram Markov Models

### **AI Application**


*   NLP

### **Business Segments**


*   Lifestyle & Social Media, Media & Publishing
*   Business & Private Sector

### **Data**

### **Papers**

### **Demo Link**








#### Importing the libraries

In [12]:
import tensorflow as tf
from tensorflow.keras.preprocessing.text import Tokenizer
import pickle
import numpy as np

#### 1 - Load The Data

In [2]:
from google.colab import files
uploaded = files.upload()

Saving pride and prejudice.txt to pride and prejudice.txt


#### 2 -Data Preprocessing, Exploratory Data Analysis

In [4]:
file = open("/content/pride and prejudice.txt" ,"r", encoding="utf8")

#store file in a list
lines = []
for i in file:
  lines.append(i)

#convert list to a string
data = ""
for i in lines:
  data = ''.join(lines)

data = data.replace('\n','').replace('\r','').replace('\ufeff','').replace('“','').replace('”','')

#remove unnecessary spaces
z = []
for i in data.split():
  if i not in z:
    z.append(i)

data = ' '.join(z)
data[:500]

'The Project Gutenberg eBook of Pride and prejudice, by Jane AustenThis is for the use anyone anywhere in United States andmost other parts world at no cost with almost restrictionswhatsoever. You may copy it, give it away or re-use under termsof License included this online atwww.gutenberg.org. If you are not located States, youwill have to check laws country where beforeusing eBook.Title: prejudiceAuthor: AustenRelease Date: November 12, 2022 [eBook #1342][Most recently updated: March 14, 2023]'

In [5]:
len(data)

217245

### **Tokenization**

In [9]:
tokenizer = Tokenizer()
tokenizer.fit_on_texts([data])

#saving the tokenizer for predict function
pickle.dump(tokenizer, open('token.pkl','wb'))

sequence_data = tokenizer.texts_to_sequences([data])[0]
sequence_data[:15]


[9, 702, 70, 408, 39, 123, 1, 409, 97, 42, 3109, 44, 35, 9, 326]

In [10]:
len(sequence_data)

25257

In [11]:
#index 0 is reserved for padding
vocab_size = len(tokenizer.word_index) + 1
print(vocab_size)

13038


We have 13038 unique words in our text corpus

In [13]:
sequences = []

for i in range(1, len(sequence_data)):
  words = sequence_data[i-1:i+1]
  sequences.append(words)

print("The length of sequences are: ",len(sequences))
sequences = np.array(sequences)
sequences[:10]

The length of sequences are:  25256


array([[   9,  702],
       [ 702,   70],
       [  70,  408],
       [ 408,   39],
       [  39,  123],
       [ 123,    1],
       [   1,  409],
       [ 409,   97],
       [  97,   42],
       [  42, 3109]])

In [14]:
#separating input and output
X =[]
y =[]

for i in sequences:
  X.append(i[0])
  y.append(i[1])

X = np.array(X)
y = np.array(y)